### ISSUE TITLE
Support arbitrary fields using Python implementation of basis functions 

### ISSUE THREAD

**Issue by leakec** [leakec/tfc#8]
## Feature
The C++ basis functions are restricted to type double. This is what is used in the majority of cases. However, there is a conjecture that TFC can be used on arbitrary fields (and its supported now with quite a bit of evidence).

One way to implement this is with C++ template classes. This would be an easy way to support complex types for example. However, there are nice Python libraries for other types, like Galois fields; of course, regular derivatives don't really make sense for these fields, so I'm not sure why we would use them with JAX derivative transformations anyway, but there could be others, e.g., single float, complex, complex single float, etc. If we can use a Python implementation of the basis functions, it would be easier to support arbitrary types.

Supporting gradients/Jacobians is easy since we can write customer derivative functions for JAX. The difficult part here is supporting JIT, since that requires a C++ implementation. However, because of #4, we no longer need JIT translations for the basis functions, since they can become compile time constants when solving differential equations. Of course, this is limiting if anyone wants JIT support for the basis functions for another reason, but I don't foresee that need at the moment.

A `backend` keyword on the `utfc` and `mtfc` classes will be used to switch from the current, C++ backend to the Python backend it desired. The default will be to use C++, for backwards compatibility, and because the overwhelming majority of cases are on the field of reals. If the user chooses the Python backend, they will gain arbitrary field specification, at the cost of the JIT translation, which again, can be overcome for differential equation or other optimization type problems if the basis functions can become compile time constants: supported using #4.

**Comment by leakec** [comment]
Note, I am swamped at work at the moment, so will not have time to work on this feature for a few months most likely. Just placing it here for now so I don't forget.

**Comment by leakec** [comment]
## To do
* [ ] Write classes for Python basis functions that mimic C++. Can make use of `pOP.py` in `tests` to make this easier.
* [ ] Add backend keyword to `utfc` and `mtfc` to switch between basis function backends. 
* [ ] Add JAX translations for Python version. Make use of #2 to support arbitrary derivatives.
* [ ] Write test cases. Should include at least one ODE, probably using complex numbers, and something from finite fields. There looks to be a nice (Galois)[https://pypi.org/project/galois/] field implementation for NumPy. Could use this plus examples from the appendix of my dissertation as a test case; likely just need to form the constrained expression here. Note, this is dependent on Galois playing nice with JAX which I haven't tested. In the worst case, the test case could use the underlying implicit primitive.
* [ ] Write a tutorial for the docs. This will be in the advanced tutorial section. I think solving a complex ODE will be sufficient. Can use the same ODE that is used in the test case.

**Comment by leakec** [comment]
Original to-do from above was:

> Write test cases. Should include at least one ODE, probably using complex numbers, and something from finite fields. There looks to be a nice [Galois](https://pypi.org/project/galois/) field implementation for NumPy. Could use this plus examples from the appendix of my dissertation as a test case; likely just need to form the constrained expression here. As mentioned before, there is really no need for the typical derivative transforms here (I think derivatives on finite fields require something different, but I'm way out of my wheelhouse here); however, we could still run it just to prove that we can do things with arbitrary types. Note, this is dependent on Galois playing nice with JAX which I haven't tested. In the worst case, the test case could use the underlying implicit primitive.

Removing the reference to Galois because this doesn't play nice with JAX. The dtype for Galois is `uint8` so we have to re-write stuff like `egrad` to play nice with `int`, by making the tangents for these types of type `[('float0', 'V')]`. It seems silly to re-write all of this for something I know doesn't make real mathematical sense, since the JAX derivatives are not designed for finite fields. I'll do the complex number example and leave it at that. That should be sufficient to show that this works.

**Comment by leakec** [comment]
Proof of concept using complex numbers, pejit, and CP basis function defined in Python:
```python
import numpy as onp
import jax.numpy as np
from jax import core, abstract_arrays
from jax.interpreters import ad, batching
from tfc.utils import pejit, egrad

deg = 5
def CP(x,d=0):
    if len(x.shape) == 1:
        x = onp.expand_dims(x,1)
    N = onp.size(x)
    One = onp.ones_like(x)
    Zero = onp.zeros_like(x)
    if deg == 0:
        if d > 0:
            F = Zero
        else:
            F = One
        return F
    elif deg == 1:
        if d > 1:
            F = onp.hstack((Zero,Zero))
        elif d > 0:
            F = onp.hstack((Zero,One))
        else:
            F = onp.hstack((One,x))
        return F
    else:
        F = onp.hstack((One,x,onp.zeros((N,deg-1),dtype=x.dtype)))
        for k in range(2,deg+1):
            F[:,k:k+1] = 2*x*F[:,k-1:k]-F[:,k-2:k-1]
        def Recurse(dark,d,dCurr=0):
            if dCurr == d:
                return dark
            else:
                if dCurr == 0:
                    dark2 = onp.hstack((Zero,One,onp.zeros((N,deg-1),dtype=x.dtype)))
                else:
                    dark2 = onp.zeros((N,deg+1),dtype=x.dtype)
                for k in range(2,deg+1):
                    dark2[:,k:k+1] = (2+2*dCurr)*dark[:,k-1:k]+2*x*dark2[:,k-1:k]-dark2[:,k-2:k-1]
                dCurr += 1
                return Recurse(dark2,d,dCurr=dCurr)
        F = Recurse(F,d)
        return F

# Create primitives
H_p = core.Primitive("H")

def Hjax(x, d=0):
    return H_p.bind(x, d=d)

# Implicit translation
def H_impl(x, d=0):
    return CP(x, d)

H_p.def_impl(H_impl)

# Abstract evaluation
def H_abstract_eval(x, d=0):
    dim1 = deg+1
    if len(x.shape) == 0:
        dims = (dim1,)
    else:
        dims = (x.shape[0], dim1)
    return abstract_arrays.ShapedArray(dims, x.dtype)

H_p.def_abstract_eval(H_abstract_eval)

# Define batching translation
def H_batch(vec, batch, d=0):
    return Hjax(*vec, d=d), batch[0]

batching.primitive_batchers[H_p] = H_batch

# Define jacobain vector product
def H_jvp(arg_vals, arg_tans, d=0, full=False):
    x = arg_vals[0]
    dx = arg_tans[0]
    if not (dx is ad.Zero):
        if type(dx) is batching.BatchTracer:
            flag = onp.any(dx.val != 0)
        else:
            flag = onp.any(dx != 0)
        if flag:
            if len(dx.shape) == 1:
                out_tans = Hjax(x, d=d + 1) * onp.expand_dims(dx, 1)
            else:
                out_tans = Hjax(x, d=d + 1) * dx
    else:
        dim0 = x.shape[0]
        dim1 = deg+1
        out_tans = np.zeros((dim0, dim1),dtype=x.dtype)
    return (Hjax(x, d=d), out_tans)

ad.primitive_jvps[H_p] = H_jvp

# Easy alias
H = Hjax

# Test out basics
xi = np.ones(deg+1)
x = onp.linspace(-1.,1.,10)
y = H(x)
z = egrad(H)(x)
g = lambda x,xi: np.dot(egrad(H)(x),xi)
gjit = pejit(x,xi,constant_arg_nums=[0])(g)
assert(np.all(onp.round(gjit(xi),10) == onp.round(g(x,xi),10)))

# Test with complex numbers
xi = np.ones(deg+1) + 1.j
x = np.linspace(-1.,1.,10)
x += 2.j
y = H(x)
z = egrad(H)(x)
g = lambda x,xi: np.dot(egrad(H)(x),xi)
gjit = pejit(x,xi,constant_arg_nums=[0])(g)
assert(np.all(onp.round(gjit(xi),10) == onp.round(g(x,xi),10)))
```e
There is no output, since the code does not throw any errors.

**Comment by leakec** [comment]
For `ELMSigmoid`, `ELMTanh`, `ELMSin`, and `ELMSwish` in the Python backend, we can use JAX gradients to extend beyond derivative order 8. In the C++ backend, the derivatives have to be entered manually, so only went up to order 8, because we did not have a need beyond that. Using the JAX gradients, we can go up to arbitrary order.

At the end, will use a `to_py` to return a regular numpy array, that way the output matches all other Python-backend basis functions.

**Comment by leakec** [comment]
Closed by merge request [#10](https://github.com/leakec/tfc/pull/10).
